src/domain/mongoose_lazy_routing.erl

```diff
@@ -0,0 +1,397 @@
+%% Generally, you should not call anything from this module directly.
+-module(mongoose_lazy_routing).
+-behaviour(gen_server).
+
+-include("mongoose_logger.hrl").
+
+%% start/stop API
+-export([start/0, stop/0]).
+-export([start_link/0]).
+
+%% domain/subdomain API
+-export([maybe_add_domain_or_subdomain/1,
+         maybe_remove_domain/2,
+         maybe_remove_subdomain/1]).
+
+%% IQ handling API
+-export([register_iq_handler_for_domain/4,
+         register_iq_handler_for_subdomain/5,
+         unregister_iq_handler_for_domain/3,
+         unregister_iq_handler_for_subdomain/4]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-define(IQ_TABLE, mongoose_lazy_routing_iqs).
+-define(ROUTING_TABLE, mongoose_lazy_routing).
+
+-record(iq_table_key, {host_type :: host_type(),
+                       %% subdomain_pattern is 'undefined' for domains
+                       subdomain_pattern :: subdomain_pattern() | undefined,
+                       namespace = '_' :: binary() | '_',
+                       component = '_' :: module() | '_'}).
+
+-type subdomain_pattern() :: mongoose_subdomain_utils:subdomain_pattern().
+-type subdomain_info() :: mongoose_subdomain_core:subdomain_info().
+-type host_type() :: mongooseim:host_type().
+-type domain() :: mongooseim:domain_name().
+
+-type iq_record() :: {#iq_table_key{}, mongoose_iq_handler:t()}.
+-type domain_record() :: {domain(), host_type() | {host_type(), subdomain_pattern()}}.
```

Comment:
Could you rename these types? The `_record` suffix is misleading, they are not records. Could be `_entry`.